src/object/cli_ec.c

```diff
@@ -33,6 +33,454 @@
 #include "obj_rpc.h"
 #include "obj_internal.h"
 
+#define EC_DEBUG 1
+
+static int
+obj_ec_recxs_init(struct obj_ec_recx_array *recxs, uint32_t recx_nr)
+{
+	if (recxs->oer_recxs != NULL) {
+		D_ERROR("oer_recxs non-NULL, cannot init again.\n");
+		return -DER_INVAL;
+	}
+
+	if (recx_nr == 0)
+		return 0;
+
+	D_ALLOC_ARRAY(recxs->oer_recxs, recx_nr);
+	if (recxs->oer_recxs == NULL)
+		return -DER_NOMEM;
+
+	return 0;
+}
+
+static void
+obj_ec_pbuf_fini(struct obj_ec_recx_array *recxs)
+{
+	int	i;
+
+	D_FREE(recxs->oer_pbufs[0]);
+	for (i = 0; i < recxs->oer_p; i++)
+		recxs->oer_pbufs[i] = NULL;
+}
+
+void
+obj_ec_recxs_fini(struct obj_ec_recx_array *recxs)
+{
+	if (recxs == NULL)
+		return;
+	if (recxs->oer_recxs != NULL)
+		D_FREE(recxs->oer_recxs);
+	recxs->oer_nr = 0;
+	recxs->oer_stripe_total = 0;
+	obj_ec_pbuf_fini(recxs);
+}
+
+static int
+obj_ec_pbufs_init(struct obj_ec_recx_array *recxs, uint64_t cell_bytes)
+{
+	void		*pbuf;
+	uint8_t		*ptmp;
+	uint64_t	 parity_len;
+	int		 i;
+
+	parity_len = roundup(recxs->oer_stripe_total * cell_bytes, 8);
+	D_ALLOC(pbuf, parity_len * recxs->oer_p);
+	if (pbuf == NULL)
+		return -DER_NOMEM;
+
+	ptmp = pbuf;
+	for (i = 0; i < recxs->oer_p; i++) {
+		recxs->oer_pbufs[i] = (void *)ptmp;
+		ptmp += parity_len;
+	}
+	return 0;
+}
+
+/* update recx_nrs on data cells */
+#define ec_all_data_tgt_recx_nrs(oca, recx_nrs, i)			       \
+	for (i = 0; i < (oca)->u.ec.e_k; i++)				       \
+		recx_nrs[i]++;
+
+/* update recx_nrs and seg_nrs for replica on parity cells */
+#define ec_replica_pairty_seg_recx(oca, recx_nrs, seg_nrs, i)		       \
+	for (i = 0; i < (oca)->u.ec.e_p; i++) {				       \
+		recx_nrs[(oca)->u.ec.e_k + i]++;			       \
+		seg_nrs[(oca)->u.ec.e_k + i]++;				       \
+	}
+
+/* update recx_nrs on all targets */
+#define ec_all_tgt_recx_nrs(oca, recx_nrs, i)				       \
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop
(style)  macros should not use a trailing semicolon